File: smoketests/tests/permissions.py
from .. import Smoketest, random_string
import json

class Permissions(Smoketest):
    AUTOPUBLISH = False

    def setUp(self):
        self.reset_config()

    def test_call(self):
        """Ensure that anyone has the permission to call any standard reducer"""

        self.publish_module()

        self.call("say_hello", anon=True)

        self.assertEqual("\n".join(self.logs(10000)).count("World"), 1)

    def test_delete(self):
        """Ensure that you cannot delete a database that you do not own"""

        self.publish_module()

        self.new_identity()
        with self.assertRaises(Exception):
            self.spacetime("delete", self.database_identity)

    def test_describe(self):
        """Ensure that anyone can describe any database"""

        self.publish_module()

        self.spacetime("describe", "--anonymous", "--json", self.database_identity)

    def test_logs(self):
        """Ensure that we are not able to view the logs of a module that we don't have permission to view"""

        self.publish_module()

        self.reset_config()
        self.new_identity()
        self.call("say_hello")

        self.reset_config()
        self.new_identity()
        with self.assertRaises(Exception):
            self.spacetime("logs", self.database_identity, "-n", "10000")

    def test_publish(self):
        """This test checks to make sure that you cannot publish to an identity that you do not own."""

        self.publish_module()

        self.new_identity()

        with self.assertRaises(Exception):
            self.spacetime("publish", self.database_identity, "--project-path", self.project_path, "--delete-data", "--yes")

        # Check that this holds without `--delete-data`, too.
        with self.assertRaises(Exception):
            self.spacetime("publish", self.database_identity, "--project-path", self.project_path, "--yes")

    def test_replace_names(self):
        """Test that you can't replace names of a database you don't own"""

        name = random_string()
        self.publish_module(name)

        self.new_identity()

        with self.assertRaises(Exception):
            self.api_call(
                "PUT",
                f'/v1/database/{name}/names',
                json.dumps(["post", "gres"]),
                {"Content-type": "application/json"}
            )

class PrivateTablePermissions(Smoketest):
    MODULE_CODE = """
use spacetimedb::{ReducerContext, Table};

#[spacetimedb::table(name = secret, private)]
pub struct Secret {
    answer: u8,
}

#[spacetimedb::table(name = common_knowledge, public)]
pub struct CommonKnowledge {
    thing: String,
}

#[spacetimedb::reducer(init)]
pub fn init(ctx: &ReducerContext) {
    ctx.db.secret().insert(Secret { answer: 42 });
}

#[spacetimedb::reducer]
pub fn do_thing(ctx: &ReducerContext, thing: String) {
    ctx.db.secret().insert(Secret { answer: 20 });
    ctx.db.common_knowledge().insert(CommonKnowledge { thing });
}
"""

    def test_private_table(self):
        """Ensure that a private table can only be queried by the database owner"""

        out = self.spacetime("sql", self.database_identity, "select * from secret")
        answer = "\n".join([
            " answer ",
            "--------",
            " 42     ",
            ""
        ])
        self.assertMultiLineEqual(str(out), answer)

        self.reset_config()
        self.new_identity()

        with self.assertRaises(Exception):
            self.spacetime("sql", self.database_identity, "select * from secret")

        # Subscribing to the private table failes.
        with self.assertRaises(Exception):
            self.subscribe("SELECT * FROM secret", n=0)

        # Subscribing to the public table works.
        sub = self.subscribe("SELECT * FROM common_knowledge", n = 1)
        self.call("do_thing", "godmorgon")
        self.assertEqual(sub(), [
            {
                'common_knowledge': {
                    'deletes': [],
                    'inserts': [{'thing': 'godmorgon'}]
                }
            }
        ])

        # Subscribing to both tables returns updates for the public one.
        sub = self.subscribe("SELECT * FROM *", n=1)
        self.call("do_thing", "howdy", anon=True)
        self.assertEqual(sub(), [
            {
                'common_knowledge': {
                    'deletes': [],
                    'inserts': [{'thing': 'howdy'}]
                }
            }
        ])


class LifecycleReducers(Smoketest):
    lifecycle_kinds = "init", "client_connected", "client_disconnected"

    MODULE_CODE = "\n".join(f"""
#[spacetimedb::reducer({kind})]
fn lifecycle_{kind}(_ctx: &spacetimedb::ReducerContext) {{}}
""" for kind in lifecycle_kinds)

    def test_lifecycle_reducers_cant_be_called(self):
        """Ensure that lifecycle reducers (init, on_connect, etc) can't be called"""

        for kind in self.lifecycle_kinds:
            with self.assertRaises(Exception):
                self.call(f"lifecycle_{kind}")

File: crates/smoketests/tests/smoketests/permissions.rs
use spacetimedb_smoketests::Smoketest;
use std::path::PathBuf;

fn workspace_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .to_path_buf()
}

/// Ensure that anyone has the permission to call any standard reducer
#[test]
fn test_call() {
    let test = Smoketest::builder().precompiled_module("modules-basic").build();

    test.call_anon("say_hello", &[]).unwrap();

    let logs = test.logs(10000).unwrap();
    let world_count = logs.iter().filter(|l| l.contains("World")).count();
    assert_eq!(world_count, 1, "Expected 1 'World' in logs, got {}", world_count);
}

/// Ensure that anyone can describe any database
#[test]
fn test_describe() {
    let test = Smoketest::builder().precompiled_module("modules-basic").build();

    // Should succeed with anonymous describe
    test.describe_anon().unwrap();
}

/// Ensure that we are not able to view the logs of a module that we don't have permission to view
#[test]
fn test_logs() {
    let test = Smoketest::builder().precompiled_module("modules-basic").build();

    // Call say_hello as owner
    test.call("say_hello", &[]).unwrap();

    // Switch to a new identity
    test.new_identity().unwrap();

    // Call say_hello as new identity (should work - reducers are public)
    test.call("say_hello", &[]).unwrap();

    // Switch to another new identity
    test.new_identity().unwrap();

    // Try to view logs - should fail as non-owner
    let identity = test.database_identity.as_ref().unwrap();
    let result = test.spacetime(&["logs", "--server", &test.server_url, identity, "-n", "10000"]);
    assert!(result.is_err(), "Expected logs to fail for non-owner");
}

/// Ensure that you cannot publish to an identity that you do not own
#[test]
fn test_publish() {
    let test = Smoketest::builder().precompiled_module("modules-basic").build();

    let identity = test.database_identity.as_ref().unwrap().clone();

    // Switch to a new identity
    test.new_identity().unwrap();

    // Try to publish with --delete-data - should fail
    let project_path = workspace_root().join("crates/smoketests/modules/modules-basic");
    let result = test.spacetime(&[
        "publish",
        &identity,
        "--server",
        &test.server_url,
        "--project-path",
        project_path.to_str().unwrap(),
        "--delete-data",
        "--yes",
    ]);
    assert!(
        result.is_err(),
        "Expected publish with --delete-data to fail for non-owner"
    );

    // Try to publish without --delete-data - should also fail
    let result = test.spacetime(&[
        "publish",
        &identity,
        "--server",
        &test.server_url,
        "--project-path",
        project_path.to_str().unwrap(),
        "--yes",
    ]);
    assert!(result.is_err(), "Expected publish to fail for non-owner");
}

/// Test that you can't replace names of a database you don't own
#[test]
fn test_replace_names() {
    let mut test = Smoketest::builder()
        .precompiled_module("modules-basic")
        .autopublish(false)
        .build();

    let name = format!("test-db-{}", std::process::id());
    test.publish_module_named(&name, false).unwrap();

    // Switch to a new identity
    test.new_identity().unwrap();

    // Try to replace names - should fail
    let json_body = r#"["post", "gres"]"#;
    let response = test
        .api_call_json("PUT", &format!("/v1/database/{}/names", name), json_body)
        .unwrap();
    assert!(
        response.status_code != 200,
        "Expected replace names to fail for non-owner, got status {}",
        response.status_code
    );
}

/// Ensure that a private table can only be queried by the database owner
#[test]
fn test_private_table() {
    let test = Smoketest::builder().precompiled_module("permissions-private").build();

    // Owner can query private table
    test.assert_sql(
        "SELECT * FROM secret",
        r#" answer
--------
 42"#,
    );

    // Switch to a new identity
    test.new_identity().unwrap();

    // Non-owner cannot query private table
    let result = test.sql("SELECT * FROM secret");
    assert!(result.is_err(), "Expected query on private table to fail for non-owner");

    // Subscribing to the private table fails
    let result = test.subscribe(&["SELECT * FROM secret"], 0);
    assert!(
        result.is_err(),
        "Expected subscribe to private table to fail for non-owner"
    );

    // Subscribing to the public table works
    let sub = test
        .subscribe_background(&["SELECT * FROM common_knowledge"], 1)
        .unwrap();
    test.call("do_thing", &["godmorgon"]).unwrap();
    let events = sub.collect().unwrap();

    assert_eq!(
        serde_json::json!(events),
        serde_json::json!([{
            "common_knowledge": {
                "deletes": [],
                "inserts": [{"thing": "godmorgon"}]
            }
        }])
    );

    // Subscribing to both tables returns updates for the public one only
    let sub = test.subscribe_background(&["SELECT * FROM *"], 1).unwrap();
    test.call("do_thing", &["howdy"]).unwrap();
    let events = sub.collect().unwrap();

    assert_eq!(
        serde_json::json!(events),
        serde_json::json!([{
            "common_knowledge": {
                "deletes": [],
                "inserts": [{"thing": "howdy"}]
            }
        }])
    );
}

/// Ensure that you cannot delete a database that you do not own
#[test]
fn test_cannot_delete_others_database() {
    let test = Smoketest::builder().build();

    let identity = test.database_identity.as_ref().unwrap().clone();

    // Switch to a new identity
    test.new_identity().unwrap();

    // Try to delete the database - should fail
    let result = test.spacetime(&["delete", "--server", &test.server_url, &identity, "--yes"]);
    assert!(result.is_err(), "Expected delete to fail for non-owner");
}

/// Ensure that lifecycle reducers (init, on_connect, etc) can't be called directly
#[test]
fn test_lifecycle_reducers_cant_be_called() {
    let test = Smoketest::builder().precompiled_module("permissions-lifecycle").build();

    let lifecycle_kinds = ["init", "client_connected", "client_disconnected"];

    for kind in lifecycle_kinds {
        let reducer_name = format!("lifecycle_{}", kind);
        let result = test.call(&reducer_name, &[]);
        assert!(
            result.is_err(),
            "Expected call to lifecycle reducer '{}' to fail",
            reducer_name
        );
    }
}
